doc/README.twom.md

```diff
@@ -0,0 +1,189 @@
+
+## The original cyrus skiplist format
+
+This format is purely just an in-memory skiplist, where new “allocations” are just appending a set of bytes to a file, and the pointers for are all relative pointers, offsets within the file.
+
+Since there is no way to “free” a pointer in this format, garbage collection is done by creating a new file and copying all the records into the new file.  This was done with a special INORDER record type, which was otherwise identical to an ADD but only occurred after a repack.
+
+## The twoskip file format
+
+Twoskip added an extra innovation to the skiplist format - a second linked list at the bottom level, which includes every single record in order - but with a twist.  The two pointers leapfrog each other, so it’s always the lowest pointer which is updated each time.  This means that if there’s a crash, the pointer which doesn’t point past the old size of the file will always point to the record which used to be the next in the valid records - so a repair is just a matter of walking all the valid level 0 pointers, and zeroing out any future pointers plus fixing up the higher level  pointers.  This is so fast that transaction `abort()` is implemented as just running recovery on the file.
+
+Along with a three phase commit, this makes the twoskip file format incredibly robust.  The three phases are:
+
+1. before writing any new record to the end of file, mark the header as DIRTY, and fsync
+2. once all new records are written, fsync the entire file (still marked dirtry)
+3. mark the header NOT-DIRTY along with the new file length, and fsync
+
+This means that so long as the final header write is atomic, transactions are always completely safe.
+
+## Twoskip downsides:
+
+- Three fsyncs per transaction is kinda expensive!
+- Every write causes the file to be extended, as twoskip format uses exact file size and doesn’t allocate slop.
+- A single write does amplify into 1+N (the level of the record) writes; the append of the new record, and the updates of the immediately previous records in the skiplist at each level.  Since half the records are level 1, half that level 2, etc - the average write updates three separate previous records.  In a recently repacked file, the previous records are likely to be adjacent or nearly so for the lower levels, but this becomes less so over time, so repacks are valuable for both data locality and size reduction.
+- The repack operation requires the database to be remain exclusively locked for the duration, which is slow on large files (particularly along with the above lack of slop).
+
+## The Twom format
+
+Why the name?  It’s twoskip, but with mmap and MVCC.  Two Ms.  And the “tombstone” records, rather than just causing the skiplist to be rewritten without the deleted record in it, remain in the linked lists right until a repack, with an ANCESTOR pointer back to the previous entity with that key.
```

Comment:
Similar to previous comments, this mentions tombstone records without more explanation.